File: client/allcoolapp/src/service/AchievementService.ts
import requestExecutor from './AxiosService';
import { AxiosPromise } from 'axios';
import { AchievementViewDTO, AchievementDTO } from '../types/dto';

const resource = '/api/achievements';

export const findById = (id: string): AxiosPromise<AchievementViewDTO> => {
  return requestExecutor.get(`${resource}/${id}/view`);
};

export const findAllAchievementByProductId = (
  productId: string
): AxiosPromise<AchievementDTO[]> => {
  return requestExecutor.get(`${resource}/${productId}`);
};

export const findAllByProductTypeId = (
  productTypeId: string
): AxiosPromise<AchievementDTO[]> => {
  return requestExecutor.get(`${resource}/product-types/${productTypeId}`);
};

export const findAllByBrandId = (
  brandId: string
): AxiosPromise<AchievementDTO[]> => {
  return requestExecutor.get(`${resource}/brands/${brandId}`);
};

export const countByProductId = (
  productId: string
): AxiosPromise<AchievementDTO> => {
  return requestExecutor.get(`${resource}/products/${productId}`);
};

export const findAllByUserId = (
  userId: string
): AxiosPromise<AchievementDTO[]> => {
  return requestExecutor.get(`${resource}/users/${userId}`);
};

File: server/src/main/java/br/com/allcool/achievement/service/AchievementService.java
package br.com.allcool.achievement.service;

import br.com.allcool.achievement.repository.AchievementRepository;
import br.com.allcool.converter.AchievementDTOConverter;
import br.com.allcool.converter.AchievementViewDTOConverter;
import br.com.allcool.dto.AchievementDTO;
import br.com.allcool.dto.AchievementViewDTO;
import br.com.allcool.exception.DataNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@Transactional(readOnly = true)
public class AchievementService {

    private final AchievementRepository repository;

    public AchievementService(AchievementRepository achievementRepository) {
        this.repository = achievementRepository;
    }

    public List<AchievementDTO> findAllAchievementByProductId(UUID productId) {

        AchievementDTOConverter converter = new AchievementDTOConverter();

        return this.repository.findAllAchievementByProductId(productId).stream()
                .map(converter::toBrandAchievement).collect(Collectors.toList());

    }

    public List<AchievementDTO> findAllByProductTypeId(UUID productTypeId) {

        AchievementDTOConverter converter = new AchievementDTOConverter();

        return this.repository.findAllByProductTypeId(productTypeId).stream()
                .map(converter::toBrandAchievement).collect(Collectors.toList());
    }

    public List<AchievementDTO> findAllByBrandId(UUID brandId) {

        AchievementDTOConverter converter = new AchievementDTOConverter();

        return this.repository.findAllByBrandId(brandId).stream()
                .map(converter::toBrandAchievement).collect(Collectors.toList());
    }

    public AchievementViewDTO findById(UUID id) {

        return new AchievementViewDTOConverter().to(this.repository.findById(id).orElseThrow(DataNotFoundException::new));

    }

    public Long countByProductId(UUID productId) {

        return repository.countByProductId(productId);

    }
}
